Doc review: Update Release 221121.7 (#1596)
SAP/spartacus-docs: _pages/home/updating-to-version-221121.7/migration.md

---
title: Update Release 221121.7
---

To update your Spartacus app to version 221121.7, you must carry out the following steps:

1. Update your Spartacus app to version 221121.5 (with Angular 19). For more information, see [Update Release 221121.5](link-added-post-conversion).

> 2. Ensure you have Node.js 22 installed. For more information, see the official [Node.js website](https://nodejs.org/en).

Reviewer: From version 221121.7, we require Node  in version ^22.22.0 (due to security constraints). Let's adjust the version mentioned in docs. Let me know if more details about why this minimal minor version is required.

Author: This is perfect. Thank you! 🙂

3. Update your Angular libraries. For more information, see [Updating Your Angular Libraries](#updating-your-angular-libraries).
4. Update Spartacus to 221121.7.

> 5. Modernize your storefront to use the standalone Bootstrap application. For more information, see [Modernizing Your Storefront to Use the Standalone Bootstrap Application](modernize-app-to-standalone-bootstrap-application.md).

Reviewer 2: ```suggestion
5. Modernize your storefront to use the standalone `bootstrapApplication()` function. For more information, see [Modernizing Your Storefront to Use the Standalone Bootstrap Application](modernize-app-to-standalone-bootstrap-application.md).
```

> 6. Optionally migrate you custom components to standalone components. For more information, see [Standalone Components in Spartacus](standalone-components-in-spartacus.md).

Reviewer: ```suggestion
6. Optionally migrate your custom components to standalone components. For more information, see [Standalone Components in Spartacus](standalone-components-in-spartacus.md).
```

Reviewer: @Platonn is this step still optional? Isn't the case that by migrating Spartacus to standalone and using `bootstrapApplication`, customers also have to migrate their components? 
Nevertheless, I think that even it technically it is possible for customers to stick with module-based components, I think the migration should be required for the sake of consistency.

Reviewer 2: > Isn't the case that by migrating Spartacus to standalone and using bootstrapApplication, customers also have to migrate their components?

When customers are "migrating to use bootstrapApplication", they will need to migrate _only_ their root component `AppComponent` to standalone (which is described in the docs of the "migration to use bootstrapApplication"). The rest of their custom components _can_ remain non-standalone.

## Updating Your Angular Libraries

Before updating Spartacus to version 221121.7, you first need to make sure your Angular libraries are up to date. Spartacus 221121.7 requires Angular 21.

You can update your application to use Angular 21 as follows:

- Start by updating Angular to version 20. Ensure that third-party dependencies are compatible with Angular 20, and verify that any breaking changes have been addressed.
- When you have finished updating to Angular 20, you can then update to Angular 21.

### Updating to Angular 20

> 1. Run the following command, which updates the Angular version locally, and also updates other third-party dependencies from the Angular ecosystme to versions that are compatible with Angular 20, such as `@ng-select/ng-select@latest`, `@ngrx/store@20`, `angular-oauth2-oidc@20`, and `ngx-infinite-scroll@latest`:

Reviewer 2: ```suggestion
1. Run the following command, which updates the Angular version locally, and also updates other third-party dependencies from the Angular ecosystem to versions that are compatible with Angular 20, such as `@ng-select/ng-select@20`, `@ngrx/store@20`, `angular-oauth2-oidc@20`, and `ngx-infinite-scroll@20`:
```

   ```bash
   ng update @angular/core@20 @angular/cli@20 @ngrx/store@20 angular-oauth2-oidc@20 @ng-select/ng-select@20 ngx-infinite-scroll@20 --force
   git add .
   git commit -m "update angular 20 and 3rd party deps angular 20 compatible"
   ```

   This command is sourced from the Angular [Update Guide](https://angular.dev/update-guide?v=19.0-20.0&l=3) for updating from version 19 to version 20.

> 2. While you are updating to Angular 20, run the migration to replace old builders located under `@angular-devkit/build-angular` with new builders located under `@angular/build`.

Reviewer 2: ```suggestion
2. While you are updating to Angular 20, you'll be asked to run the migration to replace old builders located under `@angular-devkit/build-angular` with new builders located under `@angular/build`.
```

Author: @Platonn Thank you for this suggestion. When it comes to steps, there is a "rule" that the first sentence of the step needs to contain an action. So I propose this formulation instead:

"During the update to Angular 20, when you are asked if you want to "migrate application projects to the new build system" (which is to say, replace old builders located under `@angular-devkit/build-angular` with new builders located under `@angular/build`), run this migration. If you choose not to at this step, you will also have the option to run this migration during the update to Angular 21"

Reviewer 2: I'm sorry my comment (with just a change suggestion without any context/explanation), it could be ambigious.

What I meant to address:
The old wording said "run the migration". But to my mental model (and terminology) we've already run the migration (by running `ng update`) and now we're in the process of it.
So I wanted to emphasise that _during the migration_ (that you've already run) you'll be asked about something, and than accept it.

Author: Sounds good! Thank you very much! 👍 🙂

         "type": "component"
       },
       "@schematics/angular:directive": {
         "type": "directive"
       },
       "@schematics/angular:service": {
         "type": "service"
       },
       "@schematics/angular:guard": {
         "typeSeparator": "."
       },
       "@schematics/angular:interceptor": {
         "typeSeparator": "."
       },
       "@schematics/angular:module": {
         "typeSeparator": "."
       },
       "@schematics/angular:pipe": {
         "typeSeparator": "."
       },
       "@schematics/angular:resolver": {
         "typeSeparator": "."    }
     }
   ```

   You might also be offered the option to run the following migrations when updating to Angular 20:

   ```bash
   Select the migrations that you'd like to run  
   ❯◯ [control-flow-migration] Converts the entire application to block control flow syntax.  
    ◯ [router-current-navigation] Replaces usages of the deprecated Router getCurrentNavigation method with the Router.currentNavigation signal.
   ```

   In terms of updating Spartacus to version 221121.7, these migrations are not required. However, you can safely run these migrations if you wish to.

### Updating to Angular 21

> 1. Run the following command, which updates the Angular version locally, and also updates other third-party dependencies from the Angular ecosystme to versions that are compatible with Angular 20, such as `@ng-select/ng-select@latest`, `@ngrx/store@21`, `angular-oauth2-oidc@21`, and `ngx-infinite-scroll@latest`:

Reviewer: Unfortunately, the`angular-oauth2-oidc` library is not available in version 21. We could keep it in this command with version 20 or remove it completely.

Author: Addressed in another comment (suggested commit) below. Thank you! 🙂

> 1. Run the following command, which updates the Angular version locally, and also updates other third-party dependencies from the Angular ecosystme to versions that are compatible with Angular 20, such as `@ng-select/ng-select@latest`, `@ngrx/store@21`, `angular-oauth2-oidc@21`, and `ngx-infinite-scroll@latest`:

Reviewer 2: ```suggestion
1. Run the following command, which updates the Angular version locally, and also updates other third-party dependencies from the Angular ecosystem to versions that are compatible with Angular 21, such as `@ng-select/ng-select@21`, `@ngrx/store@21`, `angular-oauth2-oidc@20`, and `ngx-infinite-scroll@21`:
```

Non-destructive hydration is an Angular feature that improves performance by reusing the server-rendered DOM instead of destroying and recreating it on the client side. This reduces the Time to Interactive (TTI) metric and provides a better user experience.

For more information, see [Hydration](https://angular.dev/guide/hydration) in the official Angular documentation.

If you are working with a module-based application, you enable non-destructive hydration by adding `provideClientHydration()` with `withEventReplay()` and `withNoHttpTransferCache()` in your `app.module.ts` file, as shown in the following example:

```ts
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import {
  provideClientHydration,
  withEventReplay,
  withNoHttpTransferCache,
} from '@angular/platform-browser';

@NgModule({
  imports: [
    BrowserModule,
    // ...
  ],
  providers: [
    provideClientHydration(withEventReplay(), withNoHttpTransferCache()),
    // ...
  ],
  // ...
})
export class AppModule { }
```

> If you are working with a standalone application, you enable non-destructive hydration by adding `provideClientHydration()` with `withEventReplay()` and `withNoHttpTransferCache()` in your `app.config.ts` file, as shown in the following example:

Reviewer 2: ```suggestion
If you are working with a standalone application (i.e. already using the `bootstrapApplication()` function), you enable non-destructive hydration by adding `provideClientHydration()` with `withEventReplay()` and `withNoHttpTransferCache()` in your `app.config.ts` file, as shown in the following example:
```

         "typeSeparator": "."    }
     }
   ```

   You might also be offered the option to run the following migrations when updating to Angular 20:

   ```bash
   Select the migrations that you'd like to run  
   ❯◯ [control-flow-migration] Converts the entire application to block control flow syntax.  
    ◯ [router-current-navigation] Replaces usages of the deprecated Router getCurrentNavigation method with the Router.currentNavigation signal.
   ```

   In terms of updating Spartacus to version 221121.7, these migrations are not required. However, you can safely run these migrations if you wish to.

### Updating to Angular 21

1. Run the following command, which updates the Angular version locally, and also updates other third-party dependencies from the Angular ecosystem to versions that are compatible with Angular 21, such as `@ng-select/ng-select@21`, `@ngrx/store@21`, `angular-oauth2-oidc@20`, and `ngx-infinite-scroll@21`:

   ```bash
   ng update @angular/core@21 @angular/cli@21 @ngrx/store@21 angular-oauth2-oidc@20 @ng-select/ng-select@21 ngx-infinite-scroll@21 --force
   git add .
   git commit -m "update angular 21 and 3rd party deps angular 21 compatible"

>    ```

Reviewer 2: The usage of v20 instead of v21 for `angular-oauth2-oidc@20` might be confusing for customers. they might create support requests suggesting it's a typo.

What about adding a disclaimer that at the time of the release, the latest available version of `angular-oauth2-oidc` is v20, but the author of the library says it works also with Angular 21 as well. Source: https://github.com/manfredsteyer/angular-oauth2-oidc/issues/1491#issuecomment-3590104924

Author: Excellent idea! And it occurs to me, we should probably update our documentation when this lib is updated. I see you guys are watching it, so if you have any news, just let me know and we can republish this doc with the updated lib version, once it becomes available 🙂

1. In `angular.json`, you can optionally remove the redundant `outputPath` property if it matches the default value.

   This is a manual change that is not handled by the Spartacus update schematics.

   In newly-created Spartacus apps that are generated with Angular 21, the `outputPath` option is skipped and implicitly defaults to `dist/<your-project-name>`. If your migrated app has the `outputPath` set to `dist/<your-project-name>`, it is recommended that you remove it from `angular.json`, since it is not necessary.

## Enabling Non-Destructive Hydration (SSR Only)

If your Spartacus app does not use server-side rendering (SSR), you do not need to enable non-destructive hydration. You can now proceed with [Modernizing Your Storefront to Use the Standalone Bootstrap Application](#modernizing-your-storefront-to-use-the-standalone-bootstrap-application).

If your storefront application does use SSR, you must enable non-destructive hydration. Enabling non-destructive hydration aligns your app with current Angular best practices, making it easier to maintain and update your storefront application.

> Non-destructive hydration is an Angular feature that improves performance by reusing the server-rendered DOM instead of destroying and recreating it on the client side. This reduces the Time to Interactive (TTI) metric and provides a better user experience.

Reviewer 2: ```suggestion
Non-destructive hydration is an Angular feature that improves performance by reusing the server-rendered DOM instead of destroying and recreating it on the client side. This helps to avoid a flickering when transitioning from server-side DOM to client-side rendered DOM in the browser, which provides a better user experience.
```

Author: Perfect 👍 🙏

Observations about this warning so far are the following:

- Since May 2025, no functional issues have been reported that are related to this warning
- Both hydration and `enabledBlocking` work correctly together in Spartacus
- During SSR, `enabledBlocking` ensures all route guards (including `CmsPageGuard`) complete before rendering
- During hydration in the browser, Angular's hydration system prevents UI flickering by reusing the server-rendered DOM

You can safely ignore this warning for now. This topic continues to be actively monitored, and if any issues are found, they will be addressed in a future release of Spartacus, as required and as part of the ongoing modernization of the SSR implementation.

## Known Issue

As described in the [Font Awesome documentation](https://docs.fontawesome.com/upgrade/whats-changed/#fixed-width-icons-by-default), the Font Awesome library introduce a breaking change in version 7.x. By default, all Font Awesome icons now display at a fixed width rather than an automatic width.

> As as result, you might see icons overlapping sibling HTML elements. So far, two occurrences have been observed, for the Product Configuration alert icon, and the legacy carousel circle icons.

Reviewer: ```suggestion
As a result, you might see icons overlapping sibling HTML elements. So far, two occurrences have been observed, for the Product Configuration alert icon, and the legacy carousel circle icons.
```

};
```

> 

Reviewer 2: I'm sorry, I realized only now this paragraph is targeted to non-existing audience.

At this time or reading the document, customers' apps are likely not yet migrated to standalone. So I guess we can remove this paragraph and code snippet.

Only in the next section below `## Modernizing Your Storefront to Use the Standalone Bootstrap Application` we'll tell customers to move this `provideClientHydration()` call from `app.module.ts` to `app.config.ts`.

Author: How about moving the entire "enabling hydration" section to the end? In that case, the text can apply to all potential audiences. Or if anything, we can remove the part that still talks about `app.module.ts`.

The `withEventReplay()` and `withNoHttpTransferCache()` options are required for the following reasons:

- `withEventReplay()` ensures that user interactions that occur before the application is fully hydrated are captured and replayed. This provides a seamless user experience even during the hydration process.
- `withNoHttpTransferCache()` disables the HTTP transfer cache for hydration. This is required for Spartacus because the storefront app uses its own state transfer mechanism. Without this option, there could be conflicts between Angular's built-in HTTP transfer cache and the custom implementation provided in Spartacus.

**Note:** Ensure you test your application thoroughly after enabling non-destructive hydration to ensure all components hydrate correctly.

You can now proceed with [Modernizing Your Storefront to Use the Standalone Bootstrap Application](#modernizing-your-storefront-to-use-the-standalone-bootstrap-application).

### Known Warning: NG05001

> After enabling hydration, you may see the following warning in your console:

Reviewer 2: ```suggestion
After enabling hydration, you may see the following warning in your console in dev-mode build:
```

- During SSR, `enabledBlocking` ensures all route guards (including `CmsPageGuard`) complete before rendering
- During hydration in the browser, Angular's hydration system prevents UI flickering by reusing the server-rendered DOM

You can safely ignore this warning for now. This topic continues to be actively monitored, and if any issues are found, they will be addressed in a future release of Spartacus, as required and as part of the ongoing modernization of the SSR implementation.

## Modernizing Your Storefront to Use the Standalone Bootstrap Application

Modern Angular 21 apps use [standalone components](https://angular.dev/reference/migrations/standalone), as well as the `bootstrapApplication()` Angular API for bootstrapping the root standalone component, instead of using the old `bootstrapModule()`.

After you have updated your Spartacus application to version 221121.7, you need to modernize your storefront app to use the Angular `bootstrapApplication()` API.

Spartacus provides dedicated schematics to automatically modernize your app to use the Angular `bootstrapApplication()` API and convert your root `AppComponent` to a standalone component.

> **Note:** The schematics to modernize your app only migrate your root `AppComponent` to become a standalone component. The schematics do not migrate your custom components to become standalone components. For more information about migrating your custom components, see [Standalone Components in Spartacus](#standalone-components-in-spartacus).

Reviewer 2: ```suggestion
**Note:** The schematics to modernize your app only migrate your root `AppComponent` to become a standalone component. The schematics do not migrate your custom components to become standalone components. For more information about migrating your custom components, see [Standalone Components in Spartacus](#standalone-components-in-spartacus). But before migrating all your custom components to standalone, please follow the migration described here below.
```

Author: Excellent catch! And as I was playing with the wording, it occurred to me, we can delete this note altogether, because the next paragraph sets up the command for running the schematics, and _right after that_, there is text saying that we strongly recommend also updating custom components to standalone components -- by implication, the schematics therefore did not take care of this task 💡 😅
